determined-ai/determined
# feat: reuse HTTP sessions

## Description




Introduce the option of using persistent HTTP sessions in `api.Session`.

Previously, each API call was wrapped in its own `requests.Session`, which meant making a new underlying TCP connection/TLS handshake every time. Sometimes this is necessary because we can't know when to close the session, and we don't want to leave too many dangling HTTP connections open. But in many cases, we make lots of API calls in quick succession within known bounds, and in these cases using a persistent HTTP session significantly improves performance:

**Reporting metrics in Core API/Detached mode:**
```python
def main():
    core_v2.init(
        defaults=core_v2.DefaultConfig(
            name="unmanaged-1-singleton",
        ),
    )

    for i in range(100):
        core_v2.train.report_training_metrics(
            steps_completed=i, metrics={"loss": random.random()}
        )

        if (i + 1) % 10 == 0:
            core_v2.train.report_validation_metrics(
                steps_completed=i, metrics={"loss": random.random()}
            )

    core_v2.close()
```
*Not using persistent HTTP sessions:* 13-14s
*With a persistent HTTP session:* 6-7s

**CLI calls that fetch multiple pages:**
i.e. `list-experiments`, `list-checkpoints`
*Before:* 0.75s
*After:* 0.3s

### `api.Session` as a context manager

This PR exposes the option of re-using the same underlying HTTP session through `api.Session` as a context manager:

```python
with api.Session(...) as sess:
  # All requests within the context share the same persistent HTTP connection.
  # Once the context exits, the session is closed.
  sess.post(...)
  sess.get(...)
```

When used directly, each HTTP call will create a new session every time. This is the current behavior, so existing code will continue to function the same way:

```python
# When called directly outside of a context, each request uses a new HTTP session.
sess = api.Session(...)
sess.post(...)
sess.get(...)
```

This functionality is implemented in Core API, where all internal API calls within a `core.Context` (`with core.init(...) as core_context:`) will now share the same persistent HTTP connection which is closed when the `core_context` context exits.

#### CLI

In the CLI, `api.Session` as a context manager is implemented as a decorator, and when used, exposes a session that is persistent within the decorated function:

```diff
- def list_experiments(args: Namespace) -> None:
-     sess = cli.setup_session(args)

+ @cli.session
+ def list_experiments(args: Namespace, sess: api.Session) -> None:
```

I only changed a few select APIs/CLIs to use the new functionality, mostly as examples of how it would be used. There are many places that would see performance benefits from persistent sessions, but I imagine that will probably be a separate effort/PR.


### Other `api.Session`refactors
Some other small refactors made to `api.Session`:
- Set auth/cert-related parameters on the session instead of passing them into each request
- `harness/determined/common/requests.py` was deleted, its logic consolidated into `harness/determined/common/api/_session.py`

## Test Plan



- Run any experiment that reports a bunch of metrics (ideally Core API / detached mode) and enable debug logs.
- Submit the experiment and the logs should show "urllib3.connectionpool: Starting new HTTP connection" only once at the start of training
- Experiment should be noticeably faster

## Commentary (optional)





## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket
